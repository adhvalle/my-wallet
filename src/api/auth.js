import { post } from './client'

export const login = () => post('/login', { username: 'vicky', password: 'secret' })
