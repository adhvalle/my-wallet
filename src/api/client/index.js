import axios from 'axios'
import toCamelcase from 'camelcase-keys'
import toSnakecase from 'snakecase-keys'
import { getRouter } from '@/router'
import { getToken } from '@/models/auth'

let client

export const getClient = () => {
  if (!client) _init()
  return client
}

export const removeHeader = key => {
  delete getClient().defaults.headers.common[key]
}

export const setHeader = (key, value) => {
  getClient().defaults.headers.common[key] = value
}

export const get = (url, query = {}, headers = {}) =>
  _request('GET', url, null, query, headers)

export const post = (url, payload = {}, headers = {}) =>
  _request('POST', url, payload, {}, headers)

const _init = () => {
  client = axios.create({ baseURL: import.meta.env.VITE_API_HOST })

  if (getToken()) client.defaults.headers.common.Authorization = getToken()

  client.interceptors.request.use(request => {
    if (request.data && request.data instanceof FormData === false) request.data = _snakecaseKeys(request.data)
    if (request.params) request.params = _snakecaseKeys(request.params)

    return request
  })

  client.interceptors.response.use(
    response => {
      if (response.data) response.data = toCamelcase(response.data, { deep: true })
      return response
    },
    error => {
      if (error?.response?.status === 401) getRouter().push({ name: 'login' })

      return Promise.reject(error)
    }
  )
}

const _snakecaseKeys = object =>
  typeof object === 'object' ? toSnakecase(object, { deep: true }) : object

const _request = async (method, url, data = {}, query = {}, headers = {}) => {
  let error = null
  const response = await getClient().request({
    method,
    url,
    data,
    params: query,
    headers
  }).catch(error => { console.warn(error) })

  return new Promise((resolve, reject) => response ? resolve(response.data) : reject(error))
}
