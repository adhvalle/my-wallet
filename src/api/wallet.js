import { get } from './client'

export const fetchWallet = () => get('/wallet')