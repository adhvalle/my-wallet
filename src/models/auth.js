import { reactive } from 'vue'
import { setHeader, removeHeader } from '@/api/client'
import { setWallet } from '@/models/wallet'
import {
  set as setStoreItem, remove as removeStoreItem, get as getStoreItem
} from '@/helpers/storage'

const state = reactive({
  token: null
})

export const getToken = () => state.token || getStoreItem('token')
export const setToken = token => {
  state.token = token
  setHeader('Authorization', token)
  token ? setStoreItem('token', token) : removeStoreItem('token')
}

export const setLoginData = data => {
  const { accessToken, tokenType } = data
  setToken(`${tokenType} ${accessToken}`)
}

export const logout = () => {
  setToken(null)
  setWallet(null)
  removeHeader('Authorization')
}
