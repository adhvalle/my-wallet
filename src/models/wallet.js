import { reactive } from 'vue'
import { fetchWallet } from '@/api/wallet'

const state = reactive({
  wallet: null
})

export const getWallet = () => state.wallet
export const setWallet = wallet => { state.wallet = wallet }

/**
 * Get or fetch a Wallet
 * @returns {Promise<Object>}
 */
export const retrieveWallet = async () => {
  if (!state.wallet) {
    const response = await fetchWallet()
    setWallet(response)
  }

  return getWallet()
}
