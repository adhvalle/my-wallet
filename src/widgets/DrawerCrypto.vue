<script setup>
import { computed, reactive, ref, watch } from 'vue'
import { sendTransaction } from '@/api/wallet'
import { getContacts, retrieveContacts, retrieveWallet } from '@/models/wallet'
import TheDrawer from '@/components/TheDrawer.vue'

const props = defineProps({
  isVisible: Boolean
})

defineEmits(['on-close'])

const myContacts = computed(getContacts)
const showDrawer = ref(false)
const formData = reactive({
  amount: '',
  currency: '',
  description: 'MyWallet transaction',
  receiver: ''
})
const currencies = ['ETH', 'DOGE', 'BTC']

const openDrawerCrypto = async () => {
  await retrieveContacts()
  showDrawer.value = true
}

const submit = async () => {
  await sendTransaction(formData)
  retrieveWallet(true)
  formData.amount = ''
  formData.currency = ''
  formData.receiver = ''
}

watch(() => props.isVisible, value => {
  value ? openDrawerCrypto() : showDrawer.value = false
})
</script>

<template>
  <TheDrawer 
    :is-visible="showDrawer" 
    @on-close="$emit('on-close')"
  >
    <pre>{{ formData }}</pre>

    <form @submit.prevent="submit">
      <select v-model="formData.receiver">
        <option
          v-for="contact in myContacts"
          :key="contact.email"
          :value="contact.email"
          v-text="contact.name"
        />
      </select>
      <div>
        <div 
          v-for="currency in currencies"
          :key="currency"
        >
          <input
            v-model="formData.currency"
            type="radio"
            name="currency"
            :key="currency"
            :id="currency"
            :value="currency"
          >
          <label 
            :for="currency" 
            v-text="currency"
          />
        </div>
      </div>
      <input 
        v-model="formData.amount"
        type="text"
      >
      <button>Send transaction</button>
    </form>
  </TheDrawer>
</template>
