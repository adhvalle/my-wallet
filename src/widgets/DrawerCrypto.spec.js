import { describe, it, expect, vi, beforeAll } from 'vitest'
import { mount, flushPromises } from '@vue/test-utils'
import { sendTransaction } from '@/api/wallet'
import { retrieveWallet } from '@/models/wallet'
import DrawerCrypto from '@/widgets/DrawerCrypto.vue'

vi.mock('@/api/wallet')
vi.mock('@/models/wallet')

describe('DrawerCrypto', () => {
  beforeAll(() => {
    retrieveWallet.mockReturnValue({})
  })
  it('renders the component', async () => {
    const wrapper = await mountComponent()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('emits on-close event', async () => {
    const wrapper = await mountComponent()
    wrapper.findComponent({ name: 'TheDrawer' }).vm.$emit('on-close')
    expect(wrapper.emitted('on-close')).toBeTruthy()
  })

  it('submits form successfully', async () => {
    sendTransaction.mockReturnValue({})
    const wrapper = await mountComponent()
    wrapper.find('.drawer-crypto__form').trigger('submit')
    await flushPromises()
    expect(wrapper.find('.drawer-crypto__response').classes()).toContain('drawer-crypto__response--success')
  })

  it('submits form with errors', async () => {
    sendTransaction.mockRejectedValue({})
    const wrapper = await mountComponent()
    wrapper.find('.drawer-crypto__form').trigger('submit')
    await flushPromises()
    expect(wrapper.find('.drawer-crypto__response').classes()).toContain('drawer-crypto__response--error')
  })

  const mountComponent = async () => {
    const wrapper = mount(DrawerCrypto)
    await flushPromises()
    return wrapper
  }
})