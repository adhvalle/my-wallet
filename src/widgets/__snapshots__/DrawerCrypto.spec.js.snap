// Vitest Snapshot v1

exports[`DrawerCrypto > renders the component 1`] = `
"<transition-stub name=\\"fade\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
  <!--v-if-->
</transition-stub>
<transition-stub name=\\"slide-x\\" appear=\\"false\\" persisted=\\"true\\" css=\\"true\\">
  <div class=\\"the-drawer\\" style=\\"display: none;\\"><a href=\\"#\\" class=\\"the-drawer__btn\\"><img src=\\"/images/icn-close.svg\\" alt=\\"close\\" class=\\"the-drawer__icn\\"></a>
    <div class=\\"drawer-crypto\\">
      <!--v-if-->
      <form class=\\"drawer-crypto__form\\"><label class=\\"drawer-crypto__label\\">Receiver</label><select class=\\"drawer-crypto__select\\"></select><label class=\\"drawer-crypto__label\\">Currency</label>
        <div class=\\"drawer-crypto__options\\">
          <div class=\\"drawer-crypto__option\\"><input class=\\"drawer-crypto__radio\\" type=\\"radio\\" name=\\"currency\\" id=\\"ETH\\" value=\\"ETH\\"><label for=\\"ETH\\">ETH</label></div>
          <div class=\\"drawer-crypto__option\\"><input class=\\"drawer-crypto__radio\\" type=\\"radio\\" name=\\"currency\\" id=\\"DOGE\\" value=\\"DOGE\\"><label for=\\"DOGE\\">DOGE</label></div>
          <div class=\\"drawer-crypto__option\\"><input class=\\"drawer-crypto__radio\\" type=\\"radio\\" name=\\"currency\\" id=\\"BTC\\" value=\\"BTC\\"><label for=\\"BTC\\">BTC</label></div>
        </div><label class=\\"drawer-crypto__label\\">Amount</label><input type=\\"text\\"><button class=\\"drawer-crypto__button\\">Send transaction</button>
      </form>
    </div>
  </div>
</transition-stub>"
`;
